ui: Enable Compilation with both web-ui and native-ui for easier binary distribution

Hi, I was in the process of fixing the packaging this for nixos, so that both the native-ui and the web-ui were working. However, when I was compiling it with both feature flags enabled I got this following error:

```
   Compiling clap v4.5.40
   Compiling yosemite v0.7.0
   Compiling emissary-util v0.4.0 (/build/source/emissary-util)
   Compiling emissary-cli v0.4.0 (/build/source/emissary-cli)
error[E0428]: the name `main` is defined multiple times
   --> emissary-cli/src/main.rs:472:1
    |
403 | fn main() -> anyhow::Result<()> {
    | ------------------------------- previous definition of the value `main` here
...
472 | fn main() -> anyhow::Result<()> {
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ `main` redefined here
    |
    = note: `main` must be defined only once in the value namespace of this module

error: native and web ui cannot be enabled at the same time
  --> emissary-cli/src/main.rs:59:1
   |
59 | compile_error!("native and web ui cannot be enabled at the same time");
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

We totally could just ship 2 binaries with the different features enabled, but I thought it would be good to see if it is part of the plan to keep the different UI's in the different binaries indefinitely. If not I threw together a quick branch that fixes this: https://github.com/cappuccinocosmico/emissary/tree/enable-both-native-and-webui-on-single-binary. If this isn't part of the project vision feel free to ignore :)

--- a/emissary-cli/src/cli.rs
+++ b/emissary-cli/src/cli.rs
@@ -171,6 +171,13 @@ pub struct RouterUiOptions {
     /// Has an effect only if web UI was enabled during compilation.
     #[arg(long, value_name = "PORT")]
     pub web_ui_port: Option<u16>,
+
+    /// Use web UI instead of native UI.
+    ///
+    /// Has an effect only if both web UI and native UI were enabled during compilation.
+    /// Defaults to native UI.
+    #[arg(long, action = clap::ArgAction::SetTrue)]
+    pub web_ui: Option<bool>,
 }
 
 #[derive(Args, Default)]

--- a/emissary-cli/src/config.rs
+++ b/emissary-cli/src/config.rs
@@ -211,6 +211,8 @@ pub struct RouterUiConfig {
     pub theme: Theme,
     pub refresh_interval: usize,
     pub port: Option<u16>,
+    #[serde(rename = "web-ui")]
+    pub web_ui: Option<bool>,
 }
 
 #[derive(Debug, Default, Clone, Serialize, Deserialize)]
@@ -334,6 +336,7 @@ impl EmissaryConfig {
                 theme: Theme::Dark,
                 refresh_interval: 5usize,
                 port: None,
+                web_ui: None,
             }),
             sam: Some(SamConfig {
                 tcp_port: 7656,
@@ -925,6 +928,7 @@ impl Config {
             theme,
             refresh_interval,
             port,
+            web_ui,
         }) = &mut self.router_ui
         {
             if let Some(selected) = arguments.router_ui.theme {
@@ -937,6 +941,10 @@ impl Config {
 
             if let Some(selected) = arguments.router_ui.web_ui_port {
                 *port = Some(selected);
+            }
+
+            if let Some(selected) = arguments.router_ui.web_ui {
+                *web_ui = Some(selected);
             }
 
             if let Some(true) = arguments.router_ui.disable_ui {
@@ -984,6 +992,7 @@ mod tests {
                 refresh_interval: None,
                 theme: None,
                 web_ui_port: None,
+                web_ui: None,
             },
             floodfill: None,
             allow_local: None,

--- a/emissary-cli/src/main.rs
+++ b/emissary-cli/src/main.rs
@@ -55,8 +55,7 @@ mod tools;
 mod tunnel;
 mod ui;
 
-#[cfg(all(feature = "native-ui", feature = "web-ui"))]
-compile_error!("native and web ui cannot be enabled at the same time");
+
 
 /// Logging target for the file.
 const LOG_TARGET: &str = "emissary";
@@ -399,7 +398,7 @@ fn main() -> anyhow::Result<()> {
     Ok(())
 }
 
-#[cfg(feature = "web-ui")]
+#[cfg(all(feature = "web-ui", not(feature = "native-ui")))]
 fn main() -> anyhow::Result<()> {
     let runtime = tokio::runtime::Runtime::new()?;
     let (shutdown_tx, shutdown_rx) = channel(1);
@@ -432,7 +431,7 @@ fn main() -> anyhow::Result<()> {
 }
 
 /// A main function which should be run within an async context.
-#[cfg(feature = "native-ui")]
+#[cfg(all(feature = "native-ui", not(feature = "web-ui")))]
 async fn native_main() -> anyhow::Result<()> {
     let (shutdown_tx, shutdown_rx) = channel(1);
     let arguments = parse_arguments().await;
@@ -468,8 +467,68 @@ async fn native_main() -> anyhow::Result<()> {
     }
 }
 
-#[cfg(feature = "native-ui")]
+#[cfg(all(feature = "native-ui", not(feature = "web-ui")))]
 fn main() -> anyhow::Result<()> {
     let runtime = tokio::runtime::Runtime::new()?;
     runtime.block_on(native_main())
+}
+
+#[cfg(all(feature = "native-ui", feature = "web-ui"))]
+fn main() -> anyhow::Result<()> {
+    let runtime = tokio::runtime::Runtime::new()?;
+    let (shutdown_tx, shutdown_rx) = channel(1);
+    let arguments = runtime.block_on(parse_arguments());
+    let RouterContext {
+        router,
+        port_mapper,
+        events,
+        router_ui_config,
+        config,
+        base_path,
+        address_book_handle,
+        router_id,
+        ..
+    } = runtime.block_on(setup_router::<TokioRuntime>(arguments))?;
+
+    let use_web_ui = router_ui_config.as_ref().and_then(|c| c.web_ui).unwrap_or(false);
+
+    if use_web_ui {
+        match router_ui_config {
+            None => {
+                runtime.block_on(router_event_loop(router, port_mapper, shutdown_rx));
+            }
+            Some(RouterUiConfig {
+                refresh_interval,
+                port,
+                ..
+            }) => {
+                runtime.spawn(async move {
+                    ui::web::RouterUi::new(events, port, refresh_interval, shutdown_tx).run().await;
+                });
+                runtime.block_on(router_event_loop(router, port_mapper, shutdown_rx));
+            }
+        }
+    } else {
+        match router_ui_config {
+            None => {
+                runtime.block_on(router_event_loop(router, port_mapper, shutdown_rx));
+            }
+            Some(RouterUiConfig { .. }) => {
+                tokio::spawn(async move {
+                    router_event_loop(router, port_mapper, shutdown_rx).await;
+                    std::process::exit(0);
+                });
+                ui::native::RouterUi::start(
+                    events,
+                    config,
+                    base_path,
+                    address_book_handle,
+                    router_id,
+                    shutdown_tx,
+                )?
+            }
+        }
+    }
+
+    Ok(())
 }

--- a/emissary-cli/src/main.rs
+++ b/emissary-cli/src/main.rs
@@ -55,8 +55,6 @@ mod tools;
 mod tunnel;
 mod ui;
 
-
-
 /// Logging target for the file.
 const LOG_TARGET: &str = "emissary";
 
@@ -474,10 +472,10 @@ fn main() -> anyhow::Result<()> {
 }
 
 #[cfg(all(feature = "native-ui", feature = "web-ui"))]
-fn main() -> anyhow::Result<()> {
-    let runtime = tokio::runtime::Runtime::new()?;
+#[tokio::main]
+async fn main() -> anyhow::Result<()> {
     let (shutdown_tx, shutdown_rx) = channel(1);
-    let arguments = runtime.block_on(parse_arguments());
+    let arguments = parse_arguments().await;
     let RouterContext {
         router,
         port_mapper,
@@ -488,30 +486,30 @@ fn main() -> anyhow::Result<()> {
         address_book_handle,
         router_id,
         ..
-    } = runtime.block_on(setup_router::<TokioRuntime>(arguments))?;
+    } = setup_router::<TokioRuntime>(arguments).await?;
 
     let use_web_ui = router_ui_config.as_ref().and_then(|c| c.web_ui).unwrap_or(false);
 
     if use_web_ui {
         match router_ui_config {
             None => {
-                runtime.block_on(router_event_loop(router, port_mapper, shutdown_rx));
+                router_event_loop(router, port_mapper, shutdown_rx).await;
             }
             Some(RouterUiConfig {
                 refresh_interval,
                 port,
                 ..
             }) => {
-                runtime.spawn(async move {
+                tokio::spawn(async move {
                     ui::web::RouterUi::new(events, port, refresh_interval, shutdown_tx).run().await;
                 });
-                runtime.block_on(router_event_loop(router, port_mapper, shutdown_rx));
+                router_event_loop(router, port_mapper, shutdown_rx).await;
             }
         }
     } else {
         match router_ui_config {
             None => {
-                runtime.block_on(router_event_loop(router, port_mapper, shutdown_rx));
+                router_event_loop(router, port_mapper, shutdown_rx).await;
             }
             Some(RouterUiConfig { .. }) => {
                 tokio::spawn(async move {

--- a/emissary-cli/src/tools/router_ui_dev.rs
+++ b/emissary-cli/src/tools/router_ui_dev.rs
@@ -149,7 +149,7 @@ pub async fn run() {
         );
     }
 
-    #[cfg(feature = "web-ui")]
+    #[cfg(all(feature = "web-ui", not(feature = "native-ui")))]
     {
         ui::web::RouterUi::new(subscriber, None, 1, shutdown_tx).run().await;
     }

